File: mac/Sources/ClaudeToolbarCore/Sessions/HooksConfig.swift
import Foundation

/// Adds or removes the toolbar's HTTP hooks in a Claude Code settings document. Other content is preserved (keys are re-sorted on write).
public enum HooksConfig {
    public enum Error: Swift.Error { case invalidJSON }

    public static let events = ["SessionStart", "UserPromptSubmit", "Notification", "Stop", "StopFailure", "SessionEnd"]
    public static let hookTimeoutSeconds = 5

    public static func hookUrl(port: Int) -> String { "http://127.0.0.1:\(port)/hook" }

    public static func isInstalled(_ json: String, url: String) throws -> Bool {
        let root = try parseObject(json)
        guard let hooks = root["hooks"] as? [String: Any] else { return false }
        return events.allSatisfy { event in
            (hooks[event] as? [[String: Any]])?.contains { hasHandler($0, url: url) } ?? false
        }
    }

    public static func install(_ json: String, url: String) throws -> String {
        var root = try parseObject(json)
        var hooks = root["hooks"] as? [String: Any] ?? [:]
        for event in events {
            var groups = hooks[event] as? [[String: Any]] ?? []
            if !groups.contains(where: { hasHandler($0, url: url) }) {
                groups.append(["hooks": [["type": "http", "url": url, "timeout": hookTimeoutSeconds]]])
            }
            hooks[event] = groups
        }
        root["hooks"] = hooks
        return try serialize(root)
    }

    public static func remove(_ json: String, url: String) throws -> String {
        var root = try parseObject(json)
        if var hooks = root["hooks"] as? [String: Any] {
            for event in events {
                guard var groups = hooks[event] as? [[String: Any]] else { continue }
                groups = groups.compactMap { group -> [String: Any]? in
                    var group = group
                    guard let handlers = group["hooks"] as? [[String: Any]] else { return group }
                    let kept = handlers.filter { !isOurs($0, url: url) }
                    if kept.isEmpty { return nil }
                    group["hooks"] = kept
                    return group
                }
                if groups.isEmpty { hooks.removeValue(forKey: event) } else { hooks[event] = groups }
            }
            if hooks.isEmpty { root.removeValue(forKey: "hooks") } else { root["hooks"] = hooks }
        }
        return try serialize(root)
    }

    private static func hasHandler(_ group: [String: Any], url: String) -> Bool {
        (group["hooks"] as? [[String: Any]])?.contains { isOurs($0, url: url) } ?? false
    }

    private static func isOurs(_ handler: [String: Any], url: String) -> Bool {
        handler["type"] as? String == "http" && handler["url"] as? String == url
    }

    private static func parseObject(_ json: String) throws -> [String: Any] {
        let trimmed = json.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return [:] }
        guard let object = try? JSONSerialization.jsonObject(with: Data(trimmed.utf8), options: [.fragmentsAllowed]),
              let root = object as? [String: Any] else { throw Error.invalidJSON }
        return root
    }

    private static func serialize(_ root: [String: Any]) throws -> String {
        let data = try JSONSerialization.data(withJSONObject: root, options: [.prettyPrinted, .sortedKeys])
        return String(decoding: data, as: UTF8.self)
    }
}

File: src/ClaudeToolbar.Core/Sessions/HooksConfig.cs
using System.Text.Json;
using System.Text.Json.Nodes;

namespace ClaudeToolbar.Core.Sessions;

/// <summary>Adds or removes the toolbar's HTTP hooks in a Claude Code settings document. Everything else in the document is preserved.</summary>
public static class HooksConfig
{
    public static readonly IReadOnlyList<string> Events =
        ["SessionStart", "UserPromptSubmit", "Notification", "Stop", "StopFailure", "SessionEnd"];

    public const int HookTimeoutSeconds = 5;

    private static readonly JsonSerializerOptions Pretty = new() { WriteIndented = true };

    public static string HookUrl(int port) => $"http://127.0.0.1:{port}/hook";

    public static bool IsInstalled(string json, string url)
    {
        var root = ParseObject(json);
        if (root["hooks"] is not JsonObject hooks) return false;
        return Events.All(ev => hooks[ev] is JsonArray groups && groups.Any(g => HasHandler(g, url)));
    }

    public static string Install(string json, string url)
    {
        var root = ParseObject(json);
        if (root["hooks"] is not JsonObject hooks)
        {
            hooks = new JsonObject();
            root["hooks"] = hooks;
        }
        foreach (var ev in Events)
        {
            if (hooks[ev] is not JsonArray groups)
            {
                groups = new JsonArray();
                hooks[ev] = groups;
            }
            if (groups.Any(g => HasHandler(g, url))) continue;
            groups.Add(new JsonObject
            {
                ["hooks"] = new JsonArray(new JsonObject { ["type"] = "http", ["url"] = url, ["timeout"] = HookTimeoutSeconds }),
            });
        }
        return root.ToJsonString(Pretty);
    }

    public static string Remove(string json, string url)
    {
        var root = ParseObject(json);
        if (root["hooks"] is JsonObject hooks)
        {
            foreach (var ev in Events)
            {
                if (hooks[ev] is not JsonArray groups) continue;
                foreach (var group in groups.OfType<JsonObject>().ToList())
                {
                    if (group["hooks"] is not JsonArray handlers) continue;
                    foreach (var handler in handlers.ToList())
                        if (IsOurs(handler, url)) handlers.Remove(handler);
                    if (handlers.Count == 0) groups.Remove(group);
                }
                if (groups.Count == 0) hooks.Remove(ev);
            }
            if (hooks.Count == 0) root.Remove("hooks");
        }
        return root.ToJsonString(Pretty);
    }

    private static bool HasHandler(JsonNode? group, string url) =>
        group is JsonObject g && g["hooks"] is JsonArray handlers && handlers.Any(h => IsOurs(h, url));

    private static bool IsOurs(JsonNode? handler, string url) =>
        handler is JsonObject o && StringOf(o["type"]) == "http" && StringOf(o["url"]) == url;

    private static string? StringOf(JsonNode? node) =>
        node is JsonValue value && value.TryGetValue<string>(out var s) ? s : null;

    /// <summary>
    /// Blank input is an empty document. Throws <see cref="JsonException"/> when the text is not a JSON object.
    /// Strict on purpose: comments and trailing commas are rejected rather than silently rewritten away,
    /// which is what the macOS build does with the same file.
    /// </summary>
    private static JsonObject ParseObject(string json)
    {
        if (string.IsNullOrWhiteSpace(json)) return new JsonObject();
        JsonNode? node;
        try
        {
            node = JsonNode.Parse(json);
        }
        catch (JsonException ex) when (ex.GetType() != typeof(JsonException))
        {
            // The reader throws its own (non-public) JsonException subclass for malformed text;
            // normalize to the base type so callers can catch a single, documented exception.
            throw new JsonException(ex.Message, ex);
        }
        return node as JsonObject ?? throw new JsonException("The settings file is not a JSON object.");
    }
}
